src/libcore/slice.rs

```diff
@@ -726,13 +710,15 @@ macro_rules! iterator {
             #[inline]
             fn next_back(&mut self) -> Option<$elem> {
                 // could be implemented with slices, but this avoids bounds checks
-                unsafe {
-                    ::intrinsics::assume(!self.ptr.is_null());
-                    ::intrinsics::assume(!self.end.is_null());
-                    if self.end == self.ptr {
-                        None
-                    } else {
-                        self.end = slice_sub_offset!(self.end, 1);
+                if self.end == self.ptr {
+                    None
+                } else {
+                    unsafe {
+                        self.end = slice_offset!(self.end, -1);
+                        if mem::size_of::<T>() != 0 {
+                            ::intrinsics::assume(!self.ptr.is_null());
+                            ::intrinsics::assume(!self.end.is_null());
```

Comment:
The previous code made the assertions prior to offsetting `self.end`. I'm assuming here that asserting after the offset instead of before it is better, because it means the call to `slice_ref!(self.end)` can then assume it's non-null.